plugins/shadow-block-converter/src/shadow_block_converter.ts

```
@@ -122,17 +143,154 @@ export class BlockShadowChange extends Blockly.Events.BlockBase {
+    // return the block as-is.
+    return shadowBlock;
+  }
+
+  // Get the parent block, and determine the connection's index in the parent.
+  let parentBlock: Blockly.Block = parentConnection.getSourceBlock();
+  const connectionIndex: number = parentBlock
+    .getConnections_(true)
+    .indexOf(parentConnection);
+
+  // Recover the state of the shadow block before it was edited. The connection
+  // should still have the original state until a new block is attached to it.
+  const originalShadowState: Blockly.serialization.blocks.State | null =
+    parentConnection.getShadowState(/* returnCurrent = */ false);
+
+  // Serialize the current state of the shadow block (after it was edited).
+  const editedBlockState: Blockly.serialization.blocks.State | null =
+    Blockly.serialization.blocks.save(shadowBlock, {
+      addCoordinates: false,
+      addInputBlocks: true,
+      addNextBlocks: true,
+      doFullSerialization: false,
+    });
+  if (originalShadowState === null || editedBlockState === null) {
+    // The serialized block states are necessary to convert the block. Without
+    // them, just return the block as-is.
+    return shadowBlock;
+  }
+
+  // If the parent block is a shadow, it must be converted first.
+  if (parentBlock.isShadow()) {
+    const newParentBlock: Blockly.Block | null =
+      reifyEditedShadowBlock(parentBlock);
+    if (newParentBlock === null) {
+      // No parent block was created, so we can't recreate the current block
+      // either.
+      return null;
+    }
+    parentBlock = newParentBlock;
+
+    // The reference to the connection is obsolete. Find it from the new parent.
+    const parentConnections: Blockly.Connection[] =
+      parentBlock.getConnections_(true);
+    if (connectionIndex == -1 || connectionIndex >= parentConnections.length) {
+      // Couldn't find the corresponding connection on the new version of the
+      // parent block.
+      return null;
+    }
```

Would it be appropriate to throw an error here instead of returning `null`?

I note that this appears to be the only place this function can return `null`, and removing `null` from the return type could simplify things a bit (including removing the need for the check of `newParentBlock` above).